lms/djangoapps/shoppingcart/models.py

```diff
@@ -824,6 +824,73 @@ def generate_pdf_invoice(self, course, course_price, quantity, sale_price):
 
         return pdf_buffer
 
+    def __unicode__(self):
+        return "company: {}".format(self.company_name)
+
+
+INVOICE_TRANSACTION_STATUSES = (
+
+    ('started', 'started'),
+    ('completed', 'completed'),
+    ('cancelled', 'cancelled'),
+)
+
+
+class InvoiceTransaction(TimeStampedModel):
+    """
+         This table capture all the information needed to support "InvoicingTransactions"
+    """
+    invoice = models.ForeignKey(Invoice)
+    amount = models.DecimalField(default=0.0, decimal_places=2, max_digits=30)
+    comments = models.TextField(null=True)
+    created_by = models.ForeignKey(User)
+    last_modified_by = models.ForeignKey(User, related_name='last_modified_by_user')
+    status = models.CharField(max_length=32, default='started', choices=INVOICE_TRANSACTION_STATUSES)
+
+
+    @classmethod
+    def add_invoice_transaction(cls, invoice_id, amount, comments, user,status):
+        """
+        This function creates a Invoice Transaction entry with payment or refund.
+        """
+        try:
+            invoice = Invoice.objects.get(id=invoice_id)
+            invoice_transaction = InvoiceTransaction(
+                invoice=invoice, amount=amount, comments=comments,
+                created_by=user, status='completed', last_modified_by=user)
+            invoice_transaction.save()
+            return invoice_transaction
+        except Invoice.DoesNotExist:
+            raise Invoice.DoesNotExist
+
+
+class InvoiceItem(TimeStampedModel):
+    """
+    This is the basic interface for invoice items.
+    invoice items are line items that fill up the invoices.
+
+    Each implementation of InvoiceItem should provide its own purchased_callback as
+    a method.
+    """
+    objects = InheritanceManager()
+    invoice = models.ForeignKey(Invoice, db_index=True)
+    qty = models.IntegerField(default=1)
+    unit_price = models.DecimalField(default=0.0, decimal_places=2, max_digits=30)
```

Comment:
We have a total field already on Invoice. Theoretically, all InvoiceItems should sum up to be the same as the number of the Invoice. Are we going to add those assertions/constraints?